### PR TITLE
[refactor] Major refactoring of old static factory code

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Description

This is a major refactoring of the static / compile-time factory code in OpenMS.
The original design of having pluggable algorithms didn't work out in practice
and led to quite a lot of code smells and violations of OOP principles.
This is an attempt to fix this and decouple major parts of OpenMS. 

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Comment by timosachsenberg** [comment]
@cbielow do you know why the (static) factory mechanism exists on OpenMS? Because of the "old" TOPP tools? If you don't see any need here I would probably try to remove it. I think it complicates things a lot and introduces strong dependencies.

**Review by cbielow — COMMENTED** [review]
I think removal is warranted.
Could we even trim down on includes now and just forward declare in some places?

Or modernize a bit by returning a unique_ptr?

**Comment by timosachsenberg** [comment]
I think in the future we could even get rid of the factory and registration.
Why having this kind of static "plugin" mechanism in place anyway?
The only place where this makes sense is in IO - and even there I think it is not really needed.

**Comment by timosachsenberg** [comment]
hmm seems I changed something regarding external/internal linkage.
OpenMS::BaseModel::registerChildren was a static function inside a class template. Now it is only a static function.
```
../src/openms/include\OpenMS/TRANSFORMATIONS/FEATUREFINDER/BaseModel_impl.h(24): error C2375: 'OpenMS::BaseModel::registerChildren': redefinition; different linkage
```
And apparently platforms differ here...

**Comment by timosachsenberg** [comment]
only thing missing is setting the GUI logger
I think the proper way here is to use runtime dependency injection to set the logger. 
But if this requires too much code changes then we can also just get away with something global make_logger and get_logger or just an "extern" variable?

File: src/openms/include/OpenMS/CONCEPT/ProgressLogger.h [hunk 1]
```diff
@@ -100,12 +104,13 @@ namespace OpenMS
...
-    /// Return the name of the factory product used for this log type
-    static String logTypeToFactoryName_(LogType type);
-
     mutable ProgressLoggerImpl* current_logger_;
 
   };
 
+  // Function pointer for injecting the GUI progress logger implementation
+  typedef ProgressLogger::ProgressLoggerImpl* (*MakeGUIProgressLoggerFunc)();
+  extern MakeGUIProgressLoggerFunc make_gui_progress_logger;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CONCEPT/ProgressLogger.h]: here I define the factory. by default it returns the NoProgressLogger. can be overwritten by OPENMS_GUI.lib to provide a Qt based logger GUIProgressLogger

File: src/openms/source/CONCEPT/ProgressLogger.cpp [hunk 2]
```diff
@@ -131,56 +115,42 @@ namespace OpenMS
 
   };
 
+  // Simple runtime plugin system for GUI progress logger.
+  // An external library (e.g., OpenMS_GUI) can set this function to provide a GUI logger.
+  // As default, it just uses the NonProgressLoggerImpl.
+  MakeGUIProgressLoggerFunc make_gui_progress_logger = 
+    []() -> ProgressLogger::ProgressLoggerImpl* { return NoProgressLoggerImpl::create(); };
```
Comment (timosachsenberg) [inline src/openms/source/CONCEPT/ProgressLogger.cpp]: here it is defaulted

File: src/openms/source/CONCEPT/ProgressLogger.cpp [hunk 3]
```diff
@@ -131,56 +115,42 @@ namespace OpenMS
...
       case CMD:
       {
-        return "CMD";
+        current_logger_ = CMDProgressLoggerImpl::create();
+        break;
       }
       case GUI:
       {
-        return "GUI";
+        current_logger_ = make_gui_progress_logger();
```
Comment (timosachsenberg) [inline src/openms/source/CONCEPT/ProgressLogger.cpp:151]: here used during runtime

File: src/openms_gui/source/VISUAL/APPLICATIONS/MISC/QApplicationTOPP.cpp [hunk 4]
```diff
@@ -35,8 +36,9 @@ namespace OpenMS
   QApplicationTOPP::QApplicationTOPP(int& argc, char** argv) :
     QApplication(argc, argv)
   {
-    // register GUI ProgressLogger that can be used in GUI tools
-    Factory<ProgressLogger::ProgressLoggerImpl>::registerProduct(GUIProgressLoggerImpl::getProductName(), &GUIProgressLoggerImpl::create);
+    // inject the GUIProgressLoggerImpl to be used by OpenMS lib via an extern variable
+    make_gui_progress_logger = 
+      []() -> ProgressLogger::ProgressLoggerImpl* { return new GUIProgressLoggerImpl(); };
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/APPLICATIONS/MISC/QApplicationTOPP.cpp:42]: here a qt implementation overwrites the extern factory method

File: src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/MapAlignmentEvaluationAlgorithm.h [hunk 5]
```diff
@@ -40,7 +40,7 @@ namespace OpenMS
     bool isSameHandle(const FeatureHandle & lhs, const FeatureHandle & rhs, const double & rt_dev, const double & mz_dev, const Peak2D::IntensityType & int_dev, const bool use_charge);
 
     /// Register all derived classes in this method
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/MapAlignmentEvaluationAlgorithm.h]: remove

File: src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/FeatureGroupingAlgorithm.h [hunk 6]
```diff
@@ -42,7 +42,7 @@ namespace OpenMS
     void transferSubelements(const std::vector<ConsensusMap> & maps, ConsensusMap & out) const;
 
     /// Register all derived classes in this method
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/MAPMATCHING/FeatureGroupingAlgorithm.h]: remove

File: src/openms/include/OpenMS/COMPARISON/CLUSTERING/ClusterFunctor.h [hunk 7]
```diff
@@ -71,7 +71,7 @@ namespace OpenMS
     virtual void operator()(DistanceMatrix<float> & original_distance, std::vector<BinaryTreeNode> & cluster_tree, const float threshold = 1) const = 0;
 
     /// registers all derived products
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/COMPARISON/CLUSTERING/ClusterFunctor.h]:
```suggestion
```

**Pull request merged**